src/groups/group.jl

```diff
@@ -0,0 +1,514 @@
+
+Note that a manifold is connected with an operation by wrapping it with
+a decorator, [`AbstractGroupManifold`](@ref). In typical cases the concrete
+wrapper [`GroupManifold`](@ref) can be used.
+"""
+abstract type AbstractGroupOperation end
+
+@doc doc"""
+    AbstractGroupManifold{<:AbstractGroupOperation} <: Manifold
+
+Abstract type for a Lie group, a group that is also a smooth manifold with a
+[smooth binary operation](@ref AbstractGroupOperation).
+`AbstractGroupManifold`s must implement at least [`inv`](@ref),
+[`identity`](@ref), [`compose_left`](@ref), and [`translate_diff`](@ref).
+Group manifolds by default forward metric-related operations to the wrapped
+manifold.
+"""
+abstract type AbstractGroupManifold{O<:AbstractGroupOperation} <: Manifold end
+
+"""
+    GroupManifold{M<:Manifold,O<:AbstractGroupOperation} <: AbstractGroupManifold{O}
+
+Decorator for a smooth manifold that equips the manifold with a group
+operation, thus making it a Lie group. See [`AbstractGroupManifold`](@ref) for
+more details.
+
+# Constructor
+
+    GroupManifold(manifold, op)
+"""
+struct GroupManifold{M<:Manifold,O<:AbstractGroupOperation} <: AbstractGroupManifold{O}
+    manifold::M
+    op::O
+end
+
+is_decorator_manifold(::GroupManifold) = Val(true)
+
+"""
+    ActionDirection
+
+Direction of action on a manifold, either [`LeftAction`](@ref) or
+[`RightAction`](@ref).
+"""
+abstract type ActionDirection end
+
+"""
+    LeftAction()
+
+Left action of a group on a manifold.
+"""
+struct LeftAction <: ActionDirection end
+
+"""
+    RightAction()
+
+Right action of a group on a manifold.
+"""
+struct RightAction <: ActionDirection end
+
+
+@doc doc"""
+    Identity(group::AbstractGroupManifold)
+
+The identity element of the group `group`.
+"""
+struct Identity{G<:AbstractGroupManifold}
+    group::G
+end
+
+(e::Identity)(x) = identity(e.group, x)
+
+@doc doc"""
+    inv(G::AbstractGroupManifold, x)
+
+Inverse $x^{-1}$ of an element $x$, such that
+$x \cdot x^{-1} = x^{-1} \cdot x = e$.
+"""
+function inv!(G::AbstractGroupManifold, y, x)
+    error("inv not implemented on $(typeof(G)) for point $(typeof(x))")
+end
+
+function inv!(G::AbstractGroupManifold, y, e::Identity)
+    indentity!(G, y)
+    return y
+end
+
+function inv(G::AbstractGroupManifold, x)
+    y = similar_result(G, inv, x)
+    inv!(G, y, x)
+    return y
+end
+
+inv(::AbstractGroupManifold, e::Identity) = e
+
+function identity!(G::AbstractGroupManifold, y, x)
+    error("identity! not implemented on $(typeof(G)) for points $(typeof(y)) and $(typeof(x))")
+end
+
+@doc doc"""
+    identity(G::AbstractGroupManifold, x)
+
+Identity element $e$, such that for any element $x$,
+$x \cdot e = e \cdot x = x$. The returned element is of a similar type to `x`.
+"""
+function identity(G::AbstractGroupManifold, x)
+    y = similar_result(G, inv, x)
+    identity!(G, y, x)
+    return y
+end
+
+"""
+    compose_left(G::AbstractGroupManifold, xs...)
```

Comment:
There could be `compose_right`. Would it be OK to call the two-argument operation just `compose` and leave `compose_left` as the function for an arbitrary number of arguments?

Reply:
The terms composition, group operation, and group law are used interchangeably to describe the binary group operation. Composition therefore takes `$\circ: g,h \to g \circ h$`. I've rarely seen a version of composition defined `$\circ: g,h \to h \circ g$`.

Left and right translation, on the other hand, _are_ commonly defined by designating one of the elements as the element being translated and the other as the element doing the translating. (left/right _actions_ often have more complicated definitions by requiring that left-action be left-associative and right-action be right-associative, a property that is trivially satisfied for translations).

So I think it makes the most sense to have a single `compose` function that is binary. But because composition is by definition associative, there's an argument to be made for permitting it to have arbitrary numbers of arguments, like the [associative scalar functions in base](https://github.com/JuliaLang/julia/blob/89a51fb98485219680c7106e94cf75fd8069f3f8/base/operators.jl#L512-L534), which was the rationale for the [`compose` function in the original PR](https://github.com/JuliaNLSolvers/Manifolds.jl/blob/43426c7c9009d6af926a382219302289188101fa/src/Group.jl#L91-L109).

Anyways, those are just my 2 cents. In general, I think we should avoid functions with explicit `_left` and `_right` in the names, since we have `LeftAction` and `RightAction`.

Reply:
> The terms composition, group operation, and group law are used interchangeably to describe the binary group operation. Composition therefore takes `$\circ: g,h \to g \circ h$`. I've rarely seen a version of composition defined `$\circ: g,h \to h \circ g$`.

In this notation `compose_left` and `compose_right` would only differ for >= 3 arguments. `compose_left(G, x, y, z) = (x \circ y) \circ z`,  `compose_left(G, x, y, z) = x \circ (y \circ z)`. Essentially left fold vs right fold. I know they should give the same answer (by associativity) but that's where the `left` came from. Moreover, I've written *should* because in some cases like the group of reparametrizations of a curve things slightly break due to discretization.

I think I'll (for now) rename `compose_left` to `compose` and leave only the binary case. Other variants can be added later.